runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -1181,6 +1248,51 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
    // Preserve null check on the unsafe object
    TR::TransformUtil::separateNullCheck(comp(), callNodeTreeTop, tracer()->debugLevel());
 
+
+   bool conversionNeeded = comp()->fe()->dataTypeForLoadOrStore(type) != type;
+
+   // If we are not able to get javaLangClass it is still inefficient to put direct Access far
+   // So in that case we will generate lowTagCmpTest to branch to indirect access if true
+   TR_OpaqueClassBlock *javaLangClass = comp()->getClassClassPointer(/* isVettedForAOT = */ true);
+   bool needNotLowTagged = javaLangClass != NULL  || conversionNeeded ;
+
+   int length;
+   const char *objTypeSig = unsafeCall->getChild(1)->getSymbolReference() ? unsafeCall->getChild(1)->getSymbolReference()->getTypeSignature(length) : NULL;
+
+   // There are three cases where we cannot be sure of the Object type at compile time:
+   // 1.) Object type info is unknown/inaccessible
+   // 2.) Object type is known at compile time to be java/lang/Object
+   // 3.) Object type is known at compile time to be some interface class (e.g.: java/lang/Cloneable, java/io/Serializable)
+   bool objTypeUnknown;
+
+   if (objTypeSig == NULL)
+      {
+      objTypeUnknown = true;
+      }
+   else
+      {
+      TR_OpaqueClassBlock *objClass = comp()->fe()->getClassFromSignature(objTypeSig, length, unsafeCall->getChild(1)->getSymbolReference()->getOwningMethod(comp()));
+      objTypeUnknown = objClass == NULL || objClass == comp()->getObjectClassPointer() || TR::Compiler->cls.isInterfaceClass(comp(), objClass);
+      }
+
+   // There are two cases where IL for type checks (array, class, lowtag) needs to be generated:
+   // 1.) Object type is unknown at compile time
+   // 2.) Object is known at compile time to be a java/lang/Class object (to match JNI implementation)
+   bool checksNeeded = objTypeUnknown || (strncmp(objTypeSig, "Ljava/lang/Class;", strlen("Ljava/lang/Class;")) == 0);
```

Comment:
doesn't need this `strncmp` anymore, since `objClass == comp()->getObjectClassPointer()` is exactly doing that already.  you can find out what is `getObjectClassPointer()` returns from source code:  it is exactly JLO.

Reply:
I believe this `strncmp` is for java/lang/Class, not java/lang/Object. Since this a special case (i.e.: requires indirect access rather than direct access), it has to be included in addition to the `objTypeUnknown` condition (which includes JLO and interface types).

That being said, since we're no longer using `strncmp` like this for JLO or the interface types, I can do the same for java/lang/Class as well, if that would be preferable.

Reply:
ah, noticed the difference.  a further question is:  you didn't do this java/lang/Class test in CAS case.  is that expected?

Reply:
You're right, I didn't do that, but that possibility definitely needs to be accounted for. I'll add that condition to the CAS case now. Thank you for catching that!